File: tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.parent = None
        self.children = []

    def add_child(self, child):
        self.children.append(child)

    def remove(self, child):
        self.children.remove(child)

    def search(self, to_search):
        found = []
        if self.value == to_search:
            found.append(self)
        for child in self.children:
            found += child.search(to_search)
        return found

    # tofind(value)
    # returns nodes that are direct children
    def find(to_find):
        for child in self.children:
            if child.value == to_find:
                return child
            return None


class Tree:
    def __init__(self, root):
        self.root = root

    # toinsert(value), parent(node/value)
    def insert(self, to_insert, parent):
        # return error if parent is not found
        to_insert_node = Node(to_insert)
        if isinstance(parent, Node):
            parent.add_child(to_insert_node)
            to_insert_node.parent = parent
        else:
            found = self.search(parent)
            parent_node = self.search(parent)[0]
            parent_node.add_child(to_insert_node)
            to_insert_node.parent = parent_node

    # toremove(node/value)
    def remove(self, to_remove):
        # return error if root is removed
        if isinstance(to_remove, Node):
            print 'GOT HERE SUCKAS'
            parent = to_remove.parent
            parent.remove(to_remove)
        else:
            to_remove_nodes = self.search(to_remove)
            for to_remove_node in to_remove_nodes:
                parent = to_remove_node.parent
                parent.remove(to_remove_node)

    # tosearch(value)
    # returns all the nodes
    def search(self, to_search):
        found = self.root.search(to_search)
        return found

    def display(self):
        queue = []
        queue.append(self.root)
        while len(queue) != 0:
            curr = queue.pop(0)
            print curr.value
            for child in curr.children:
                queue.append(child)

# Test
root = Node('7')
tree = Tree(root)
root = tree.root
tree.insert('3', root)
tree.insert('11', root)
tree.insert('7', '3')
tree.insert('5', '3')
tree.insert('4', '5')
tree.insert('10', '11')
tree.insert('15', '11')
tree.insert('8', '11')
tree.insert('9', '15')
tree.display()

print 'removal of 11'
seven = tree.root.children[1]
tree.remove('5')
tree.display()

File: tree.js
function Node(value) {
    this.value = value;
    this.parent = null;
    this.children = [];

    // adds a node to the node's children
    this.add_child = function(child) {
        this.children.push(child);
    };

    // removes a node from the node's children
    this.remove = function(child) {
        var i = this.children.indexOf(child);
        this.children.splice(i, 1);
    };

    // accepts a value, returns an array of nodes
    this.search = function(to_search) {
        var found = [];
        if (this.value === to_search) {
            found.push(this);
        }
        for (var i=0; i < this.children.length; i++) {
            var child = this.children[i];
            found = found.concat(child.search(to_search));
        }
        return found;
    };

    // accepts a value, returns nodes that are direct children
    this.find = function(to_find) {
        for (var i=0; i < this.children.length; i++) {
            var child = this.children[i];
            if (child.value === to_find) {
                return child
            }
        }
        return null;
    };
}

function Tree(root) {
    this.root = root;

    // to_insert(node/value), parent(node/value)
    this.insert = function(to_insert, parent) {
        if (!(to_insert instanceof Node)) {
            var to_insert_node = new Node(to_insert);
        } else {
            var to_insert_node = to_insert;
        }
        if (parent instanceof Node) {
            parent.add_child(to_insert_node);
            to_insert_node.parent = parent;
        } else {
            var parent_node = this.search(parent)[0];
            if (parent_node === undefined) {
                throw new Error('Specified parent not found.');
            }
            parent_node.add_child(to_insert_node);
            to_insert_node.parent = parent_node;
        }
    }

    // remove(node/value)
    this.remove = function(to_remove) {
        if (this.root === null) {
            throw new Error('There is nothing to remove.');
        }
        if (to_remove instanceof Node) {
            if (to_remove !== this.root) {
                var parent = to_remove.parent;
                parent.remove(to_remove); 
            } else {
                this.root = null;
            }
        } else {
            if (to_remove !== this.root.value) {
                var to_remove_nodes = this.search(to_remove);
                if (to_remove_nodes.length === 0) {
                    throw new Error('The specified item does not exist.');
                }
                for (var i=0; i < to_remove_nodes.length; i++) {
                    var to_remove_node = to_remove_nodes[i];
                    var parent = to_remove_node.parent;
                    parent.remove(to_remove_node);
                }
            } else {
                this.root = null;
            }
        }
    }

    // tosearch(value), returns all the nodes
    this.search = function(to_search) {
        var found = [];
        if (this.root !== null) {
            found = this.root.search(to_search);
        }
        return found;
    }

    this.display = function() {
        var queue = [];
        var tree_str = '';
        if (this.root !== null) {
            tree_str = 'Current Tree <br>';
            var curr_level_count = 1;
            var next_level_count = 0;
            queue.push(this.root);
            while (queue.length !== 0) {
                var curr = queue.shift();
                tree_str += curr.value + ' ';
                curr_level_count --;
                for (var i=0; i < curr.children.length; i++) {
                    var child = curr.children[i];
                    queue.push(child);
                }
                next_level_count += curr.children.length;
                if (curr_level_count === 0) {
                    tree_str += '<br>';
                    curr_level_count = next_level_count;
                    next_level_count = 0;
                }
            }
            return tree_str;
        } else {
            return 'The tree is empty.';
        }
    }
}